Avoid unnecessary CPU-to-GPU copy of torch.load with CUDA

lw opened the pull request:
Summary:
When `torch.load` needs to load a tensor, no matter which device it will be end up being loaded on, it first creates a CPU storage for it of the necessary size. This storage is allocated but it's not "set" yet, hence no data is written to it: it exists in the kernel's memory map, but it's not resident and doesn't take up physical pages. Then, this storage is passed to the `map_location` function (if the parameter is a string, a device or a map, PyTorch builds that function automatically). The default map for CUDA consists effectively in `lambda storage, _: storage.cuda()` (I omitted the code needed to pick the correct device). This creates a GPU storage and copies over the data of the CPU storage. *This step is unnecessary as we're copying uninitialized memory*. (Surprisingly enough, though, it appears the kernel is smart enough that reading from the unpaged CPU memory doesn't cause it to become paged.) Once `map_location` returns a storage residing on the correct target device, `torch.load` resumes reading the file and copying the tensor's content over into the storage. This will overwrite the content that had previously been written to it, which confirms that the above copy was pointless.

A way to avoid this useless copy is to just create and return a new empty storage on the target GPU, instead of "transforming" the original one.

This does indeed increase the performance:
```
In [5]: torch.save(torch.rand(100, 100, 100), "/tmp/tensor")

In [6]: %timeit torch.load("/tmp/tensor", map_location="cuda")
1.55 ms ± 111 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [7]: %timeit torch.load("/tmp/tensor", map_location=lambda storage, _: torch.cuda.FloatStorage(storage.size()))
1.03 ms ± 44 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Credit for this diff is shared with Adam Lerer and Francisco Massa.

Differential Revision: D13897354

apaszke reviewed (changes requested):
The description of the PR applies only to the current serialization format, but will cause the legacy path to return storages with uninitialized memory.

lw committed:
Avoid unnecessary CPU-to-GPU copy of torch.load with CUDA (#16619)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/16619

When `torch.load` needs to load a tensor, no matter which device it will be end up being loaded on, it first creates a CPU storage for it of the necessary size. This storage is allocated but it's not "set" yet, hence no data is written to it: it exists in the kernel's memory map, but it's not resident and doesn't take up physical pages. Then, this storage is passed to the `map_location` function (if the parameter is a string, a device or a map, PyTorch builds that function automatically). The default map for CUDA consists effectively in `lambda storage, _: storage.cuda()` (I omitted the code needed to pick the correct device). This creates a GPU storage and copies over the data of the CPU storage. *This step is unnecessary as we're copying uninitialized memory*. (Surprisingly enough, though, it appears the kernel is smart enough that reading from the unpaged CPU memory doesn't cause it to become paged.) Once `map_location` returns a storage residing on the correct target device, `torch.load` resumes reading the file and copying the tensor's content over into the storage. This will overwrite the content that had previously been written to it, which confirms that the above copy was pointless.

A way to avoid this useless copy is to just create and return a new empty storage on the target GPU, instead of "transforming" the original one.

This does indeed increase the performance:
```
In [5]: torch.save(torch.rand(100, 100, 100), "/tmp/tensor")

In [6]: %timeit torch.load("/tmp/tensor", map_location="cuda")
1.55 ms ± 111 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [7]: %timeit torch.load("/tmp/tensor", map_location=lambda storage, _: torch.cuda.FloatStorage(storage.size()))
1.03 ms ± 44 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Credit for this diff is shared with Adam Lerer and Francisco Massa.

Differential Revision: D13897354

fbshipit-source-id: 2be962e161a165a46a6adf6d81341b81eac287a9

lerks commented:
Changed it to improve performance for the current format while keeping compatibility with the legacy one, in a rather hacky way. If you want this done more properly I'll need guidance on what this might mean.

ezyang commented:
Yeah, we definitely want to take this fix, but I agree that sticking the attribute on storage is a bit hacky. @apaszke, do you have any ideas about how to structure this better? We've kind of shot ourselves in the foot with the API here, because `map_location` is all about moving the storage one way or another.

Maybe there's a way to do the original version of this patch, and then fix up legacy to use another codepath which does a copy. Because there was a force push, I can't see the original version of the patch, but I imagine it unconditionally ignored the previous data? That seems like the right thing to do, assuming that you're not doing the legacy path. I think it's also acceptable to make the legacy path slower, if it simplifies the code.

apaszke reviewed (approved):
Not the most beautiful piece of code I've seen, but LGTM

apaszke commented:
But I also agree that it might be nicer to change those methods to allocate the storage on the right device and fix up the legacy paths (and it's ok to make them slower).

lerks commented:
For the record, I tried to change the `_cpu_deserialize` and `_cuda_deserialize` method to take a third boolean parameter `copy_data`, but this was breaking some tests. (I think it's because this also required changing the `register_package` function that, although barely documented, is technically a public interface so probably some users are relying on this and were broken by my change). I didn't take that deep a look to see whether this was indeed the case and how it could be worked around.

lerks commented:
I re-published this as [D14147673](https://phabricator.internmc.facebook.com/D14147673).
